### ISSUE TITLE
Add a 'data' option to forest() function

### ISSUE THREAD

**Issue by befriendabacterium** [wviechtb/metafor#60]
## Classification: 

Feature request

## Summary

forest() function currently does not have a 'data =' option to specify the dataframe where the yi and vi variables come from, as an alternative to specifying them as as individual objects in the environment. Would be handy to have, as in escalc() etc.

**Comment by wviechtb** [comment]
This is something that I would have liked (and tried) to implement on several occasions. The problem is that ``forest()`` is a generic that does S3 dispatching via its first argument. When using something like ``forest(yi, vi, data=dat)``, one would have to go through the process of evaluating ``yi`` in the context of ``dat`` first before one can do the dispatching and this raises various issues. There was some discussion around this on R-help (https://stat.ethz.ch/pipermail/r-help/2021-September/472046.html) and some possible ways to handle this, but these came with caveats. For now, I have decided that this just isn't something that can be easily done.

In fact, for this very reason, I changed several functions (like ``regtest()`` and ``ranktest()``) that used to be generics and that used dispatching to just 'plain' functions so all the logic of checking the class of the first argument could be handled within that single function and a ''data'' argument could be easily added. Doing the same for the various forest functions (i.e., converting it to a single function) would be super tricky at this point, because there are a few inconsistencies in how these functions work.

The relatively easy solution is to use ``with(dat, forest(yi, vi))`` for dispatching to ``forest.default()`` and this essentially accomplishes the same thing as if there was a ``data`` argument.
